### ISSUE TITLE
Lower performance using the official checkpoint

### ISSUE THREAD

**Issue by gaobb** [xiaozhen228/VCP-CLIP#3]
hi, @xiaozhen228

The re-produced performance is lower than those reported in your paper on multiple metrics when I test on the VisA using your released checkpoint ([train_mvtec.pth](https://drive.google.com/file/d/1eKuE7FbyGXgs7n2nKlGWD1tDcu69wt4a/view?usp=drive_link)).  The detailed comparisons are shown below.

Could you please explain the possible reason behind this GAP? Or can you use the same checkpoint to report the results?

Thank you so much, I look forward to hearing from you.

|    |   auroc_px |   aupro_px |   ap_px   |   auroc_sp  |   ap_sp | 
|:-----------|-----------:|-----------:|--------:|--------:|------:|
| paper      |      95.7 |     **90.7**  |   **30.1** |      **83.8**  |   **87.6** |  
| re-produced       |      **95.78** |      89.45 |   28.98 |      81.59  |   86.09 |

**Comment by xiaozhen228** [comment]
> hi, @xiaozhen228
> 
> The re-produced performance is lower than those reported in your paper on multiple metrics when I test on the VisA using your released checkpoint ([train_mvtec.pth](https://drive.google.com/file/d/1eKuE7FbyGXgs7n2nKlGWD1tDcu69wt4a/view?usp=drive_link)). The detailed comparisons are shown below.
> 
> Could you please explain the possible reason behind this GAP? Or can you use the same checkpoint to report the results?
> 
> Thank you so much, I look forward to hearing from you.
> 
> auroc_px	aupro_px	ap_px	auroc_sp	ap_sp
> paper	95.7	**90.7**	**30.1**	**83.8**	**87.6**
> re-produced	**95.78**	89.45	28.98	81.59	86.09

Thanks for your question. We will address your concerns from the following three aspects:

* In the paper, we report the average results of multiple experiments, which explains why the results from a single test may differ slightly from those in the paper. The uploaded train_mvtec.pth is the result of training with a random seed of 333, and you can obtain this weight by running [train.sh](https://github.com/xiaozhen228/VCP-CLIP/blob/main/train.sh) as well.

* Due to the relatively simple distribution of images in the MVTec-AD dataset, which primarily consists of symmetrical structures, using it as an auxiliary training set may lead to unstable training, causing the model to overfit the auxiliary dataset and become more sensitive to the random seed. Furthermore, the anomalous regions in the VisA dataset are generally smaller, making cross-product transfer more challenging.

* We have provided another pre-trained weight using MVTec-AD as the auxiliary training set, which achieves better performance compared to the previously provided one. Please try it at your convenience. [train_mvtec.pth](https://drive.google.com/file/d/1uJE25wx2OgSbVPMhR2rbOO5ey0r9UkTr/view?usp=sharing)

We hope our response can alleviate your concerns.

**Comment by gaobb** [comment]
hi, @xiaozhen228 

Thank you very much for your detailed explanation. 

The current re-production does not involve re-training, but directly uses your released checkpoint. I understand that the results reported in your paper are based on the average of 3 experiments, so they are different from a single experiment.

However, my previous concerns were not fully addressed. Based on your response, I summarize them as follows, and I hope you can take care of them when you have time.

1.  Is my re-produced result the same as yours when using the same checkpoint? 

2. The qualitative segmentation results (Fig. 5) should correspond to a certain checkpoint, what is the corresponding performance?

3. The performance of the new checkpoint you provide is better than the previous one. What is the specific performance? Why is the performance better than the previous version? What are the main changes?

Best Regards,
Bin-Bin


> > hi, @xiaozhen228
> > The re-produced performance is lower than those reported in your paper on multiple metrics when I test on the VisA using your released checkpoint ([train_mvtec.pth](https://drive.google.com/file/d/1eKuE7FbyGXgs7n2nKlGWD1tDcu69wt4a/view?usp=drive_link)). The detailed comparisons are shown below.
> > Could you please explain the possible reason behind this GAP? Or can you use the same checkpoint to report the results?
> > Thank you so much, I look forward to hearing from you.
> > auroc_px	aupro_px	ap_px	auroc_sp	ap_sp
> > paper	95.7	**90.7**	**30.1**	**83.8**	**87.6**
> > re-produced	**95.78**	89.45	28.98	81.59	86.09
> 
> Thanks for your question. We will address your concerns from the following three aspects:
> 
> * In the paper, we report the average results of multiple experiments, which explains why the results from a single test may differ slightly from those in the paper. The uploaded train_mvtec.pth is the result of training with a random seed of 333, and you can obtain this weight by running [train.sh](https://github.com/xiaozhen228/VCP-CLIP/blob/main/train.sh) as well.
> * Due to the relatively simple distribution of images in the MVTec-AD dataset, which primarily consists of symmetrical structures, using it as an auxiliary training set may lead to unstable training, causing the model to overfit the auxiliary dataset and become more sensitive to the random seed. Furthermore, the anomalous regions in the VisA dataset are generally smaller, making cross-product transfer more challenging.
> * We have provided another pre-trained weight using MVTec-AD as the auxiliary training set, which achieves better performance compared to the previously provided one. Please try it at your convenience. [train_mvtec.pth](https://drive.google.com/file/d/1uJE25wx2OgSbVPMhR2rbOO5ey0r9UkTr/view?usp=sharing)
> 
> We hope our response can alleviate your concerns.

**Comment by xiaozhen228** [comment]
> hi, @xiaozhen228
> 
> Thank you very much for your detailed explanation.
> 
> The current re-production does not involve re-training, but directly uses your released checkpoint. I understand that the results reported in your paper are based on the average of 3 experiments, so they are different from a single experiment.
> 
> However, my previous concerns were not fully addressed. Based on your response, I summarize them as follows, and I hope you can take care of them when you have time.
> 
> 1. Is my re-produced result the same as yours when using the same checkpoint and this repo?
> 2. The qualitative segmentation results (Fig. 5) should correspond to a certain checkpoint, what is the corresponding performance?
> 3. The performance of the new checkpoint you provide is better than the previous one. What is the specific performance? Why is the performance better than the previous version? What are the main changes?
> 
> Best Regards, Bin-Bin
> 
> > > hi, @xiaozhen228
> > > The re-produced performance is lower than those reported in your paper on multiple metrics when I test on the VisA using your released checkpoint ([train_mvtec.pth](https://drive.google.com/file/d/1eKuE7FbyGXgs7n2nKlGWD1tDcu69wt4a/view?usp=drive_link)). The detailed comparisons are shown below.
> > > Could you please explain the possible reason behind this GAP? Or can you use the same checkpoint to report the results?
> > > Thank you so much, I look forward to hearing from you.
> > > auroc_px	aupro_px	ap_px	auroc_sp	ap_sp
> > > paper	95.7	**90.7**	**30.1**	**83.8**	**87.6**
> > > re-produced	**95.78**	89.45	28.98	81.59	86.09
> > 
> > 
> > Thanks for your question. We will address your concerns from the following three aspects:
> > 
> > * In the paper, we report the average results of multiple experiments, which explains why the results from a single test may differ slightly from those in the paper. The uploaded train_mvtec.pth is the result of training with a random seed of 333, and you can obtain this weight by running [train.sh](https://github.com/xiaozhen228/VCP-CLIP/blob/main/train.sh) as well.
> > * Due to the relatively simple distribution of images in the MVTec-AD dataset, which primarily consists of symmetrical structures, using it as an auxiliary training set may lead to unstable training, causing the model to overfit the auxiliary dataset and become more sensitive to the random seed. Furthermore, the anomalous regions in the VisA dataset are generally smaller, making cross-product transfer more challenging.
> > * We have provided another pre-trained weight using MVTec-AD as the auxiliary training set, which achieves better performance compared to the previously provided one. Please try it at your convenience. [train_mvtec.pth](https://drive.google.com/file/d/1uJE25wx2OgSbVPMhR2rbOO5ey0r9UkTr/view?usp=sharing)
> > 
> > We hope our response can alleviate your concerns.

Thanks for your questions.
**Q1:** Is my re-produced result the same as yours when using the same checkpoint and this repo?
**A1:** Yes, the previously uploaded checkpoint will produce the same results as those obtained after your re-production.

**Q3:** The performance of the new checkpoint you provide is better than the previous one. What is the specific performance? Why is the performance better than the previous version? What are the main changes?
**A3:** The results obtained using the newly provided [train_mvtec.pth](https://drive.google.com/file/d/1uJE25wx2OgSbVPMhR2rbOO5ey0r9UkTr/view?usp=sharing) are shown in Table A. We did not make any additional changes compared to the previous training code, except for using a different random seed. Why does the random seed have such a significant impact on model performance? We believe that the training process is inherently a supervised learning task, which can easily cause the model to overfit to the auxiliary training dataset after a certain training epoch. Therefore, the initialization of parameters can affect when the model achieves better generalization performance during fine-tuning. In the paper, we uniformly selected the 10th epoch for convenience, but in fact, for MVTec-AD, the model may perform better at the 8th epoch. The issue related to prompt learning has also been pointed out in previous works [R1]. We will attempt to address this problem and upload pretrained weights with different random seeds in the future.

[R1] Ma C, Liu Y, Deng J, et al. Understanding and mitigating overfitting in prompt tuning for vision-language models[J]. IEEE Transactions on Circuits and Systems for Video Technology, 2023, 33(9): 4616-4629.

> Table A

|   AUROC_px   |   AUPRO_px   |   AP_px   |   AUROC_sp   |   AP_sp   |
|:------------:|:------------:|:---------:|:------------:|:---------:|
|    95.76     |    90.41     |   29.85   |    83.87     |   87.62   |


**Q2**: The qualitative segmentation results (Fig. 5) should correspond to a certain checkpoint, what is the corresponding performance?

**A2**: For the visualization results in Fig. 5, we most likely used the pretrained weights from [train_mvtec.pth](https://drive.google.com/file/d/1uJE25wx2OgSbVPMhR2rbOO5ey0r9UkTr/view?usp=sharing), and the performance is shown in Table A of **A3**. Additionally, for the images selected in Table 5, using different pretrained weights has little impact on the visualization results, as they are relatively simple. The main challenge for zero-shot anomaly segmentation using supervised prompt learning lies in logically anomalous products, such as PCBs.

We hope our response can alleviate your concerns.

**Comment by gaobb** [comment]
hi，@xiaozhen228

Thank you very much for your detailed explanation. My issues have been well addressed.


> > hi, @xiaozhen228
> > Thank you very much for your detailed explanation.
> > The current re-production does not involve re-training, but directly uses your released checkpoint. I understand that the results reported in your paper are based on the average of 3 experiments, so they are different from a single experiment.
> > However, my previous concerns were not fully addressed. Based on your response, I summarize them as follows, and I hope you can take care of them when you have time.
> > 
> > 1. Is my re-produced result the same as yours when using the same checkpoint and this repo?
> > 2. The qualitative segmentation results (Fig. 5) should correspond to a certain checkpoint, what is the corresponding performance?
> > 3. The performance of the new checkpoint you provide is better than the previous one. What is the specific performance? Why is the performance better than the previous version? What are the main changes?
> > 
> > Best Regards, Bin-Bin
> > > > hi, @xiaozhen228
> > > > The re-produced performance is lower than those reported in your paper on multiple metrics when I test on the VisA using your released checkpoint ([train_mvtec.pth](https://drive.google.com/file/d/1eKuE7FbyGXgs7n2nKlGWD1tDcu69wt4a/view?usp=drive_link)). The detailed comparisons are shown below.
> > > > Could you please explain the possible reason behind this GAP? Or can you use the same checkpoint to report the results?
> > > > Thank you so much, I look forward to hearing from you.
> > > > auroc_px	aupro_px	ap_px	auroc_sp	ap_sp
> > > > paper	95.7	**90.7**	**30.1**	**83.8**	**87.6**
> > > > re-produced	**95.78**	89.45	28.98	81.59	86.09
> > > 
> > > 
> > > Thanks for your question. We will address your concerns from the following three aspects:
> > > 
> > > * In the paper, we report the average results of multiple experiments, which explains why the results from a single test may differ slightly from those in the paper. The uploaded train_mvtec.pth is the result of training with a random seed of 333, and you can obtain this weight by running [train.sh](https://github.com/xiaozhen228/VCP-CLIP/blob/main/train.sh) as well.
> > > * Due to the relatively simple distribution of images in the MVTec-AD dataset, which primarily consists of symmetrical structures, using it as an auxiliary training set may lead to unstable training, causing the model to overfit the auxiliary dataset and become more sensitive to the random seed. Furthermore, the anomalous regions in the VisA dataset are generally smaller, making cross-product transfer more challenging.
> > > * We have provided another pre-trained weight using MVTec-AD as the auxiliary training set, which achieves better performance compared to the previously provided one. Please try it at your convenience. [train_mvtec.pth](https://drive.google.com/file/d/1uJE25wx2OgSbVPMhR2rbOO5ey0r9UkTr/view?usp=sharing)
> > > 
> > > We hope our response can alleviate your concerns.
> 
> Thanks for your questions. **Q1:** Is my re-produced result the same as yours when using the same checkpoint and this repo? **A1:** Yes, the previously uploaded checkpoint will produce the same results as those obtained after your re-production.
> 
> **Q3:** The performance of the new checkpoint you provide is better than the previous one. What is the specific performance? Why is the performance better than the previous version? What are the main changes? **A3:** The results obtained using the newly provided [train_mvtec.pth](https://drive.google.com/file/d/1uJE25wx2OgSbVPMhR2rbOO5ey0r9UkTr/view?usp=sharing) are shown in Table A. We did not make any additional changes compared to the previous training code, except for using a different random seed. Why does the random seed have such a significant impact on model performance? We believe that the training process is inherently a supervised learning task, which can easily cause the model to overfit to the auxiliary training dataset after a certain training epoch. Therefore, the initialization of parameters can affect when the model achieves better generalization performance during fine-tuning. In the paper, we uniformly selected the 10th epoch for convenience, but in fact, for MVTec-AD, the model may perform better at the 8th epoch. The issue related to prompt learning has also been pointed out in previous works [R1]. We will attempt to address this problem and upload pretrained weights with different random seeds in the future.
> 
> [R1] Ma C, Liu Y, Deng J, et al. Understanding and mitigating overfitting in prompt tuning for vision-language models[J]. IEEE Transactions on Circuits and Systems for Video Technology, 2023, 33(9): 4616-4629.
> 
> > Table A
> 
> AUROC_px	AUPRO_px	AP_px	AUROC_sp	AP_sp
> 95.76	90.41	29.85	83.87	87.62
> **Q2**: The qualitative segmentation results (Fig. 5) should correspond to a certain checkpoint, what is the corresponding performance?
> 
> **A2**: For the visualization results in Fig. 5, we most likely used the pretrained weights from [train_mvtec.pth](https://drive.google.com/file/d/1uJE25wx2OgSbVPMhR2rbOO5ey0r9UkTr/view?usp=sharing), and the performance is shown in Table A of **A3**. Additionally, for the images selected in Table 5, using different pretrained weights has little impact on the visualization results, as they are relatively simple. The main challenge for zero-shot anomaly segmentation using supervised prompt learning lies in logically anomalous products, such as PCBs.
> 
> We hope our response can alleviate your concerns.
